rclcpp/include/rclcpp/wait_set_policies/dynamic_storage.hpp

```diff
@@ -216,6 +216,11 @@ class DynamicStorage : public rclcpp::wait_set_policies::detail::StoragePolicyCo
       shared_waitables_
     );
 
+    if(this->needs_pruning_) {
+      this->storage_prune_deleted_entities();
+      this->needs_pruning_ = false;
+    }
+
     this->storage_release_ownerships();
```

Comment:
Should this check and the pruning happen after releasing ownerships? It seems like holding the ownership by the storage might cause the weak pointers to have not expired even though all the user's shared pointers have been destroyed? Maybe it gets caught on the next run through?

I suppose this would only happen if something lost all ownership (except the wait set's) after `needs_pruning_` became true but before this step.

Either way, since `storage_prune_deleted_entities()` iterates through all of the wait set's weak pointers and checks them for expired, we might as well release ownerships first no?

Reply:
@wjwwood does this answer your question ?